libraries/chain/apply_context.cpp

```diff
@@ -303,7 +303,7 @@ void apply_context::execute_deferred( deferred_transaction&& trx ) {
    } FC_CAPTURE_AND_RETHROW((trx));
 }
 
-void apply_context::cancel_deferred( uint128_t sender_id ) {
+void apply_context::cancel_deferred( const uint128_t& sender_id ) {
```

Comment:
Sorry I didn't think of this earlier, but why don't we make deferred_reference (and everything else) take a transaction_id_type and then we don't need to cast the hash to an uint128_t?

Reply:
For non-generated deferred transactions, the user knows the sender id as they are the ones that set it, so they can cancel them easily.  But, for their deferred transactions they would have to create table entries and track deferred transactions for each deferred they send.  With leaving it as sender id, they can use an account or code, account or receiver, account pair and not have to track them.

Reply:
I'm saying, why have a deferred_tansaction have a sender and sender_id (which is a portion of the transaction_id), why not have sender and transaction_id, and then a generated_transaction_object just stores the sender and transaction_id (and not the sender_id, which is just redundant information from the transaction_id).  Basically everywhere we have just a sender_id, we instead have a transaction_id, and everywhere we have both, we just have the transaction_id.

Reply:
I get that and agree with your statement, but we only have redundant information for a subset of all deferred transactions.  And I still think that allowing the contract writer the ability to write their own generator for sender id or use there account is more beneficial than any confusion it might raise.